### PR TITLE
Add support for QLOG JSON-SEQ

### PR DESCRIPTION
**Pull request by LPardue**
Introduces the `QlogFileSeq` and `TraceSeq` types defined in draft-ietf-quic-qlog-main-schema-01, and changes the qlog version to `0.3`. Default the streaming serializer to JSON-SEQ format.

### CODE REVIEW

File: tools/qlog/src/streamer.rs [hunk 1]
```diff
@@ -74,56 +80,105 @@ impl QlogStreamer {
...
+        }
+
+        if let Some(ref qlog_seq) = self.qlog_seq {
+            // The `QlogSeq` contains a simple `TraceSeq`, so we can write
+            // it out directly as a standalone item.
+
+            match serde_json::to_string(qlog_seq) {
+                Ok(out) => {
+                    let out = format!("
{}\n", out);
```
[the remark this answers is not in the document]
Comment (ghedo) [reply tools/qlog/src/streamer.rs:174]: In alternative I also just realized it would probably be even better to replace all the `serde_json::to_string()` calls with `serde_json::to_writer()`, which would avoid additional temporary allocations and make things more efficient https://docs.rs/serde_json/1.0.69/serde_json/fn.to_writer.html

It can probably be left for a future refactoring though, so we can either do the `write!()` change I suggested above now, and `to_writer()` later, or just do `to_writer()` now, but that's probably slightly trickier as it changes the code structure a bit.

**Review by ghedo — CHANGES_REQUESTED** [review]

File: tools/qlog/src/streamer.rs [hunk 2]
```diff
@@ -54,10 +54,16 @@ pub enum StreamerState {
     Finished,
 }
 
+pub enum SerializationFormat {
```
[the remark this answers is not in the document]
Comment (LPardue) [reply tools/qlog/src/streamer.rs:57]: This is a good question. This PR is currently a kitchen sink approach. There's a couple of points to pick at here.

1. We should IMO support both `Trace` and `TraceSeq` in qlog/src/lib.rs. Both of these formats will exist in the qlog specs for the foreseeable. As you correctly point out, this aids deserializers for both formats (which I tested in separate code). 

2a. Keeping `Trace` for serialization purposes is basically a no-op wrt serde. Applications can use the object directly and handle their own serialization and `write()`s themselves, no obligation to use streamer.rs. If they  can live with buffering, things are easy. If they can't or don't want to buffer, the code for streaming writes isn't too hard to come up with but..

2b. streamer.rs is a weird thing anyway. It adds a streaming serializing mechanism for an inherently non-streaming logging format. This helps avoid costly buffering but seems to be working against the design intentions. 

This PR represents the "kitchen sink" approach and doesn't seem to add too much complexity over the hoops the streaming serializer already had to jump through. However, if we accept that streamer.rs made an initial misstep, then there is a good opportunity to simplify the code. Depending on how much we are willing the modify the current behaviour, the following options are available

* streamer.s supports only JSON-SEQ format - avoids internal branching 
* events are only written atomically - simplifies internal state management and makes the streamer.rs usage a bit more friendly
  * in other words, drop the ability to start event, write frames, end event. Instead all events need to be written with their contents. This would require applications to buffer a bit more than they do today. The worst case is a single QUIC packets with lots of frames. That doesn't seem too onerous to me.
  * the upside of making events more atomic is that we avoid problems today when event writes quietly fail because the serializer is stuck in "frame-writing mode". That is an annoying problem we encountered in the quiche library.

My preference is to do both of these things even if it breaks the current API slightly.

File: tools/qlog/src/streamer.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (LPardue) [reply tools/qlog/src/streamer.rs]: This all becomes easier if we just write out qlog events atomically. Building on top of this branch, here is another branch that would change to atomic events that are written with `serde_json::to_writer()` - https://github.com/cloudflare/quiche/compare/lucas/json-seq...lucas/atomic-json-events?expand=1

File: tools/qlog/src/streamer.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (LPardue) [reply tools/qlog/src/streamer.rs]: PR updated to only serialize to JSON-SEQ. The atomic event thing is punted to follow up work.

**Comment by LPardue** [comment]
@ghedo regarding your earlier comment about minimizing temporary allocations, we can actually do this without too much effort and without changing the API. I pushed https://github.com/cloudflare/quiche/pull/1080/commits/c5d6759a7ac41d13f9af92457ff58ec9c2eea54f to show that

The only outlier remains events that contain frames (packet sent and received events). Because of the way we allow frame appends in the API, we need the string version and to trim the end off. I don't think it's feasible to write!() the event directly because all the serialization magic is buried inside of serde autogen code. If we do go the atomic event route - the serialization code would be simplified down to a single serde_json::with_writer() call, as can be seen on https://github.com/cloudflare/quiche/compare/lucas/json-seq...lucas/atomic-json-events?expand=1

**Review by ghedo — APPROVED** [review]

**Comment by LPardue** [comment]
the last push was just a rebase on master. Please take another look @ghedo

**Review by ghedo — APPROVED** [review]
